inmoose/cohort_qc/cohort_metric.py

```diff
@@ -0,0 +1,864 @@
+
+            data_before = self.data_expression_df_before[selected_samples]
+            data_before_melted = data_before.melt(
+                var_name="Sample", value_name="Expression"
+            )
+            data_before_melted["Dataset"] = self.clinical_df.loc[
+                data_before_melted["Sample"], self.batch_column
+            ].values
+            data_before_melted["Type"] = data_before_melted["Dataset"].apply(
+                lambda x: "Mixed" if x in self.mixed_datasets else "Non-Mixed"
+            )
+            data_before_melted["Dataset"] = data_before_melted["Dataset"].apply(
+                truncate_name, 11
+            )
+            # Plot before correction
+            sns.violinplot(
+                ax=axs[0],
+                x="Dataset",
+                y="Expression",
+                hue="Type",
+                data=data_before_melted,
+                palette={"Mixed": "red", "Non-Mixed": "grey"},
+            )
+            axs[0].set_title("Gene expression distribution before correction")
+            axs[0].set_xlabel("Dataset (Batch)")
+            axs[0].set_ylabel("Gene Expression")
+            axs[0].tick_params(axis="x", rotation=45)
+            axs[0].legend(title="Dataset Type", loc="upper right")
+            ax = axs[1]
+        else:
+            fig, ax = plt.subplots(nrows=1, ncols=1, figsize=(fig_width, 5))
+
+        # Plot after correction
+        sns.violinplot(
+            ax=ax,
+            x="Dataset",
+            y="Expression",
+            hue="Type",
+            data=data_after_melted,
+            palette={"Mixed": "red", "Non-Mixed": "grey"},
+        )
+        ax.set_title("Gene expression distribution after correction")
+        ax.set_xlabel("Dataset (Batch)")
+        ax.set_ylabel("Gene Expression")
+        ax.tick_params(axis="x", rotation=45)
+        ax.legend(title="Dataset Type", loc="upper right")
+
+        # Return the plot as a figure object
+        fig.tight_layout()
+        return fig
+
+    def compare_sample_distribution_total(self) -> plt.Figure:
+        """
+        Compare the overall distribution of samples by dataset before and after correction across all covariate combinations.
+
+        Returns
+        -------
+        matplotlib.figure.Figure
+            The generated plot object.
+        """
+        self.clinical_df["Covariate_Combination"] = (
+            self.clinical_df[self.covariates].astype(str).agg("-".join, axis=1)
+        )
+
+        data_after = self.data_expression_df.copy()
+        mean_after = data_after.mean(axis=0).reset_index()
+        mean_after.columns = ["Sample", "Mean_Expression"]
+        mean_after["Covariate_Combination"] = self.clinical_df.loc[
+            mean_after["Sample"], "Covariate_Combination"
+        ].values
+
+        width = (11 / (12 - min(11, mean_after["Covariate_Combination"].nunique()))) + 4
+        heigth = max(
+            10,
```

Comment:
Where do all these constants come from?

Reply:
I've just done some tests to make sure the plot isn't too small when there are lots of datasets. Do you think it will be better to set these values as parameters?

Reply:
No they'll be to difficult to explain, if you are happy with the tests it's good :)